### PR TITLE
element is leaf and forest is local checks [2/3]

### PR DESCRIPTION
**Pull request by holke**
**_Describe your changes here:_**

- Added a function `t8_forest_element_is_leaf` to check whether a given element is a leaf in a forest.
- Added a function `t8_forest_tree_is_local` to check whether a given tree id belongs to a local tree.
- Added a test for t8_forest_element_is_leaf

Merge after #1079 

**_All these boxes must be checked by the reviewers before merging the pull request:_**

As a reviewer please read through all the code lines and make sure that the code is fully understood, bug free, well-documented and well-structured.

#### General
- [ ] The reviewer executed the new code features at least once and checked the results manually

- [ ] The code follows the [t8code coding guidelines](https://github.com/holke/t8code/wiki/Coding-Guideline)
- [ ] New source/header files are properly added to the Makefiles
- [ ] The code is well documented
- [ ] All function declarations, structs/classes and their members have a proper doxygen documentation
- [ ] All new algorithms and data structures are sufficiently optimal in terms of memory and runtime (If this should be merged, but there is still potential for optimization, create a new issue)

#### Tests
- [ ] The code is covered in an existing or new test case using Google Test

#### Github action

- [ ] The code compiles without warning in debugging and release mode, with and without MPI (this should be executed automatically in a github action)
- [ ] All tests pass (in various configurations, this should be executed automatically in a github action)

  If the Pull request introduces code that is not covered by the github action (for example coupling with a new library):
  - [ ] Should this use case be added to the github action?
  - [ ] If not, does the specific use case compile and all tests pass (check manually)

#### Scripts and Wiki

- [ ] If a new directory with source-files is added, it must be covered by the `script/find_all_source_files.scp` to check the indentation of these files.
- [ ] If this PR introduces a new feature, it must be covered in an example/tutorial and a Wiki article.

#### Licence

- [ ] The author added a BSD statement to `doc/` (or already has one)

### CODE REVIEW

File: src/t8_forest/t8_forest_cxx.cxx [hunk 1]
```diff
@@ -1723,8 +1723,8 @@ t8_forest_leaf_face_neighbors_ext (t8_forest_t forest, t8_locidx_t ltreeid, cons
   int num_children_at_face, at_maxlevel;
   int ineigh, *owners, different_owners, have_ghosts;
 
-  /* TODO: implement is_leaf check to apply to leaf */
   T8_ASSERT (t8_forest_is_committed (forest));
+  T8_ASSERT (t8_forest_is_leaf (leaf));
```
Comment (lukasdreyer) [inline src/t8_forest/t8_forest_cxx.cxx:1727]:
```suggestion
  T8_ASSERT (t8_forest_element_is_leaf (forest, leaf, ltreeid);
```

File: src/t8_forest/t8_forest_general.h [hunk 2]
```diff
@@ -493,6 +504,19 @@ t8_forest_cmesh_ltreeid_to_ltreeid (t8_forest_t forest, t8_locidx_t lctreeid);
...
 t8_forest_get_coarse_tree (t8_forest_t forest, t8_locidx_t ltreeid);
 
+/**
+ * Query whether a given element is a leaf in a forest.
+ * 
+ * \param [in]  forest    The forest.
+ * \param [in]  element   An element of a local tree in \a forest.
+ * \param [in]  local_tree A local tree id of \a forest.
+ * \return True (non-zero) if and only if \a element is a leaf in \a local_tree of \a forest.
+ * \note This does not query for ghost leafs.
```
Comment (lukasdreyer) [inline src/t8_forest/t8_forest_general.h:514]:
```suggestion
 * \note This does not query for ghost leaves.
```

File: src/t8_forest/t8_forest_cxx.cxx [hunk 3]
```diff
@@ -2019,6 +2019,48 @@ t8_forest_print_all_leaf_neighbors (t8_forest_t forest)
...
+  if (search_result < 0) {
+    /* The element was not found. */
+    return 0;
+  }
+  /* An element was found but it may not be the candidate element. 
+   * To identify whether the element was found, we compare these two. */
+  const t8_element_t *check_element = t8_element_array_index_locidx (elements, search_result);
+  T8_ASSERT (check_element != NULL);
+  /* If the compare function returns 0, the elements are equal and we return true. */
+  return (scheme->t8_element_compare (element, check_element) == 0);
```
Comment (lukasdreyer) [inline src/t8_forest/t8_forest_cxx.cxx:2061]:
```suggestion
  return (scheme->t8_element_equal (element, check_element));
```
Please do not use compare, if you only want to check for equality. Compare uses the linearid, which is much more expensive than just a check for equality! Also, it impedes readability.

File: test/t8_forest/t8_gtest_element_is_leaf.cxx [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+{
+  t8_test_element_is_leaf_for_forest (forest);
+}
+
+TEST_P (element_is_leaf, element_is_leaf_adapt)
+{
+  t8_test_element_is_leaf_for_forest (forest_adapt);
+}
+
+/* TODO: Figure out how to do pretty printing for the cmesh examples when they are in a combined parameter. */
```
Comment (lukasdreyer) [inline test/t8_forest/t8_gtest_element_is_leaf.cxx:149]: What is the status of this TODO, is it helpful here, or better in a location for all cmeshes?

Comment (Davknapp) [reply test/t8_forest/t8_gtest_element_is_leaf.cxx:149]: Something like this 
```suggestion
auto pretty_print_params = [] (const testing::TestParamInfo<std::tuple<int, cmesh_example_base *>> &info) {
  std::string name = std::string ("Level_") + std::to_string (std::get<0> (info.param));
  std::string cmesh_name;
  std::get<1> (info.param)->param_to_string (cmesh_name);
  name += std::string ("_") + cmesh_name;
  return name;
};
```
should solve it

Comment (holke) [reply test/t8_forest/t8_gtest_element_is_leaf.cxx]: Thank you @Davknapp for the suggestion!

**Review by lukasdreyer — APPROVED** [review]

**Pull request merged**